File: server/lib/claudeSummary.ts
/**
 * Claude Opus summaries via Databricks Model Serving.
 * Default AI path for tab narratives; Supervisor Agent is optional via SUMMARY_PROVIDER=supervisor.
 */
import type { MetricsPayload } from '../../shared/types/dashboard.js';
import { buildFilterContext, buildMetricsGroundingContext, type SummaryEntity } from './summaryPrompts.js';

type ChatMessage = { role: 'system' | 'user' | 'assistant'; content: string };

const PROFESSIONAL_VOICE = `You are a senior manufacturing operations analyst writing an executive briefing for plant leadership.
Focus ONLY on unplanned downtime. Use ONLY numbers from the dashboard metrics block — never invent values.
Write 2–4 polished sentences in professional business English (not a robotic data dump).
Lead with the most important insight (performance vs prior period or benchmark).
Name specific sites, periods, categories, lines, or reasons from the data when relevant.
End with one clear, actionable recommendation for plant managers.
Do not use bullet points, markdown, headers, or JSON in the narrative text.
Vary sentence openings — avoid starting every summary with "Unplanned DT % is".`;

const TAB_FOCUS: Record<SummaryEntity, string> = {
  overview: 'Overall unplanned DT performance: KPI level, peak/low periods, STOPS, and one priority action.',
  category: 'Top downtime categories by site and period — which category drives the most loss.',
  line: 'Worst line-level contributors and site/line combinations to address first.',
  dow: 'Day-of-week and shift patterns — when unplanned DT concentrates across weeks.',
  reason: 'Top root-cause reasons by hours and period trend — what to fix first.',
};

function host(): string {
  return (process.env.DATABRICKS_HOST || process.env.DATABRICKS_SERVER_HOSTNAME || '')
    .replace(/^https?:\/\//, '')
    .replace(/\/$/, '');
}

function token(): string {
  return process.env.DATABRICKS_PAT_TOKEN || process.env.DATABRICKS_TOKEN || '';
}

/** Default Databricks Claude serving endpoint name. */
export function claudeEndpoint(): string {
  return (
    process.env.CLAUDE_SERVING_ENDPOINT
    || process.env.DATABRICKS_CLAUDE_ENDPOINT
    || 'databricks-claude-opus-4-6'
  );
}

export function claudeConfigured(): boolean {
  return Boolean(host() && token() && claudeEndpoint());
}

function claudeApiMode(): 'invocations' | 'anthropic' | 'auto' {
  const mode = String(process.env.CLAUDE_API_MODE || 'auto').toLowerCase();
  if (mode === 'anthropic' || mode === 'messages') return 'anthropic';
  if (mode === 'invocations' || mode === 'chat') return 'invocations';
  return 'auto';
}

export function isRoboticTemplateSummary(text: string): boolean {
  return /^Unplanned DT % is \d/.test(String(text || '').trim());
}

export function claudeStatus(): Record<string, unknown> {
  return {
    configured: claudeConfigured(),
    host: host() || 'NOT SET',
    endpoint: claudeEndpoint(),
    api_mode: claudeApiMode(),
  };
}

function topEntries(obj: Record<string, number[] | number>, n = 5): Record<string, unknown> {
  if (!obj) return {};
  const entries = Object.entries(obj);
  if (typeof entries[0]?.[1] === 'number') {
    return Object.fromEntries(
      entries.sort((a, b) => Number(b[1]) - Number(a[1])).slice(0, n),
    );
  }
  return Object.fromEntries(entries.slice(0, n));
}

export function compactMetrics(metrics: MetricsPayload, entityType?: SummaryEntity): Record<string, unknown> {
  const base: Record<string, unknown> = {
    kpis: metrics.kpis,
    periods: metrics.periods,
    period_trend: metrics.period_trend,
    top_reasons: (metrics.reasons || []).slice(0, 5),
    top_lines: topEntries(metrics.top_lines || {}, 5),
    shift_comparison: (metrics.shift_comparison || []).slice(0, 3),
  };

  if (!entityType || entityType === 'overview') {
    base.top_sites = topEntries(metrics.site_by_period || {}, 5);
  }
  if (!entityType || entityType === 'category') {
    base.categories = topEntries(metrics.category_by_period || {}, 5);
  }
  if (!entityType || entityType === 'line') {
    base.lines = topEntries(metrics.line_by_period || {}, 5);
  }
  if (!entityType || entityType === 'dow') {
    base.dow_sample = Object.fromEntries(
      Object.entries(metrics.dow_by_day_week || {}).slice(0, 3),
    );
  }

  return base;
}

function extractText(data: Record<string, unknown>): string {
  const choices = data.choices as Array<{ message?: { content?: string } }> | undefined;
  if (choices?.[0]?.message?.content) return choices[0].message.content.trim();

  if (typeof data.output === 'string') return data.output.trim();

  if (Array.isArray(data.output)) {
    for (const block of data.output) {
      if (block && typeof block === 'object') {
        const b = block as Record<string, unknown>;
        if (b.type === 'message' && Array.isArray(b.content)) {
          for (const cb of b.content) {
            if (cb && typeof cb === 'object') {
              const t = String((cb as Record<string, unknown>).text || '').trim();
              if (t) return t;
            }
          }
        }
      }
    }
  }

  const predictions = data.predictions as unknown[] | undefined;
  if (predictions?.[0]) {
    const p = predictions[0];
    if (typeof p === 'string') return p.trim();
    if (p && typeof p === 'object') {
      const po = p as Record<string, unknown>;
      if (typeof po.content === 'string') return po.content.trim();
      if (Array.isArray(po.candidates)) {
        const c = po.candidates[0] as Record<string, unknown> | undefined;
        if (typeof c?.text === 'string') return c.text.trim();
      }
    }
  }

  return '';
}

function parseNarrative(raw: string): string {
  const text = raw.trim();
  try {
    const d = JSON.parse(text) as Record<string, unknown>;
    if (typeof d.narrative === 'string') return d.narrative.trim();
  } catch {
    // plain text
  }
  const cleaned = text.replace(/```json\s*|```\s*/g, '').trim();
  try {
    const d = JSON.parse(cleaned) as Record<string, unknown>;
    if (typeof d.narrative === 'string') return d.narrative.trim();
  } catch {
    // plain text
  }
  return cleaned.slice(0, 600);
}

function parseBatchNarratives(raw: string): Partial<Record<SummaryEntity, string>> {
  const text = raw.replace(/```json\s*|```\s*/g, '').trim();
  try {
    const d = JSON.parse(text) as Record<string, unknown>;
    const out: Partial<Record<SummaryEntity, string>> = {};
    for (const tab of ['overview', 'category', 'line', 'dow', 'reason'] as SummaryEntity[]) {
      const v = d[tab];
      if (typeof v === 'string' && v.trim()) out[tab] = v.trim();
      else if (v && typeof v === 'object' && typeof (v as Record<string, unknown>).narrative === 'string') {
        out[tab] = String((v as Record<string, unknown>).narrative).trim();
      }
    }
    if (Object.keys(out).length) return out;
  } catch {
    // fall through
  }
  return {};
}

async function invokeClaudeInvocations(messages: ChatMessage[]): Promise<string> {
  const h = host();
  const url = `https://${h}/serving-endpoints/${claudeEndpoint()}/invocations`;
  const body = {
    messages,
    max_tokens: Number(process.env.CLAUDE_MAX_TOKENS || 800),
    temperature: Number(process.env.CLAUDE_TEMPERATURE || 0.2),
  };

  console.log(`[claude] invocations → ${claudeEndpoint()} (${messages.length} messages)…`);
  const started = Date.now();

  const resp = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token()}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });

  if (!resp.ok) {
    const errText = await resp.text();
    throw new Error(`Claude invocations HTTP ${resp.status}: ${errText.slice(0, 400)}`);
  }

  const data = (await resp.json()) as Record<string, unknown>;
  const text = extractText(data);
  if (!text) throw new Error('Claude invocations returned empty content');
  console.log(`[claude] ✓ invocations ${text.length} chars in ${Date.now() - started}ms`);
  return text;
}

async function invokeClaudeAnthropic(messages: ChatMessage[]): Promise<string> {
  const h = host();
  const url = `https://${h}/serving-endpoints/anthropic/v1/messages`;
  const system = messages.find(m => m.role === 'system')?.content;
  const chatMessages = messages
    .filter(m => m.role !== 'system')
    .map(m => ({ role: m.role, content: m.content }));

  const body: Record<string, unknown> = {
    model: claudeEndpoint(),
    max_tokens: Number(process.env.CLAUDE_MAX_TOKENS || 800),
    temperature: Number(process.env.CLAUDE_TEMPERATURE || 0.2),
    messages: chatMessages.length ? chatMessages : [{ role: 'user', content: messages[messages.length - 1]?.content || '' }],
  };
  if (system) body.system = system;

  console.log(`[claude] anthropic/messages → ${claudeEndpoint()}…`);
  const started = Date.now();

  const resp = await fetch(url, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token()}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });

  if (!resp.ok) {
    const errText = await resp.text();
    throw new Error(`Claude anthropic HTTP ${resp.status}: ${errText.slice(0, 400)}`);
  }

  const data = (await resp.json()) as Record<string, unknown>;
  const blocks = data.content as Array<{ type?: string; text?: string }> | undefined;
  const text = blocks?.map(b => b.text || '').join('').trim() || extractText(data);
  if (!text) throw new Error('Claude anthropic returned empty content');
  console.log(`[claude] ✓ anthropic ${text.length} chars in ${Date.now() - started}ms`);
  return text;
}

export async function invokeClaude(messages: ChatMessage[]): Promise<string> {
  const mode = claudeApiMode();
  if (mode === 'anthropic') return invokeClaudeAnthropic(messages);
  if (mode === 'invocations') return invokeClaudeInvocations(messages);

  try {
    return await invokeClaudeInvocations(messages);
  } catch (first) {
    console.warn(`[claude] invocations failed (${(first as Error).message.slice(0, 120)}) — trying anthropic/messages`);
    return invokeClaudeAnthropic(messages);
  }
}

export async function getClaudeTabSummary(
  entityType: SummaryEntity,
  filters: Record<string, unknown>,
  metrics: MetricsPayload,
): Promise<{ narrative: string; source: 'claude' }> {
  const filterCtx = buildFilterContext(filters);
  const grounding = buildMetricsGroundingContext(metrics, entityType);

  const user = `${PROFESSIONAL_VOICE}

Tab focus: ${TAB_FOCUS[entityType]}
Applied filters: ${filterCtx}

${grounding}

Return JSON only: {"narrative":"your 2-4 sentence executive summary"}`;

  const raw = await invokeClaude([
    { role: 'user', content: user },
  ]);

  const narrative = parseNarrative(raw);
  if (!narrative) throw new Error('Claude returned empty narrative');
  return { narrative, source: 'claude' };
}

/** Single Claude call for all KPI tabs — faster than 5 separate invocations. */
export async function getClaudeBatchSummaries(
  filters: Record<string, unknown>,
  metrics: MetricsPayload,
): Promise<Record<SummaryEntity, string>> {
  const filterCtx = buildFilterContext(filters);
  const grounding = buildMetricsGroundingContext(metrics, 'overview');

  const user = `${PROFESSIONAL_VOICE}

Applied filters: ${filterCtx}

${grounding}

Additional metrics JSON (categories, lines, reasons, shifts):
${JSON.stringify(compactMetrics(metrics), null, 2)}

Write a distinct executive summary for each KPI tab below.
Return ONLY valid JSON with these keys (each value is a 2-4 sentence string):
{
  "overview": "...",
  "category": "...",
  "line": "...",
  "dow": "...",
  "reason": "..."
}`;

  const raw = await invokeClaude([
    { role: 'user', content: user },
  ]);

  const parsed = parseBatchNarratives(raw);
  const tabs: SummaryEntity[] = ['overview', 'category', 'line', 'dow', 'reason'];
  const out = {} as Record<SummaryEntity, string>;

  for (const tab of tabs) {
    if (parsed[tab]) {
      out[tab] = parsed[tab]!;
    } else {
      const single = await getClaudeTabSummary(tab, filters, metrics);
      out[tab] = single.narrative;
    }
  }

  return out;
}

File: py_server/lib/claude_summary.py
"""
Claude Opus summaries via Databricks Model Serving.
Default AI path for tab narratives; Supervisor Agent is optional via SUMMARY_PROVIDER=supervisor.
"""
from __future__ import annotations

import json
import os
import re
import time
from typing import Any, Literal

from py_server.lib.databricks_fetch import databricks_fetch, databricks_host, databricks_token
from py_server.lib.summary_prompts import (
    SummaryEntity,
    build_filter_context,
    build_metrics_grounding_context,
)

ChatMessage = dict[str, str]

PROFESSIONAL_VOICE = """You are a senior manufacturing operations analyst writing an executive briefing for plant leadership.
Focus ONLY on unplanned downtime. Use ONLY numbers from the dashboard metrics block — never invent values.
Write 2–4 polished sentences in professional business English (not a robotic data dump).
Lead with the most important insight (performance vs prior period or benchmark).
Name specific sites, periods, categories, lines, or reasons from the data when relevant.
End with one clear, actionable recommendation for plant managers.
Do not use bullet points, markdown, headers, or JSON in the narrative text.
Vary sentence openings — avoid starting every summary with "Unplanned DT % is"."""

TAB_FOCUS: dict[SummaryEntity, str] = {
    'overview': 'Overall unplanned DT performance: KPI level, peak/low periods, STOPS, and one priority action.',
    'category': 'Top downtime categories by site and period — which category drives the most loss.',
    'line': 'Worst line-level contributors and site/line combinations to address first.',
    'dow': 'Day-of-week and shift patterns — when unplanned DT concentrates across weeks.',
    'reason': 'Top root-cause reasons by hours and period trend — what to fix first.',
}


def claude_endpoint() -> str:
    return (
        os.getenv('CLAUDE_SERVING_ENDPOINT')
        or os.getenv('DATABRICKS_CLAUDE_ENDPOINT')
        or 'databricks-claude-opus-4-6'
    )


def claude_configured() -> bool:
    return bool(databricks_host() and databricks_token() and claude_endpoint())


def _claude_api_mode() -> Literal['invocations', 'anthropic', 'auto']:
    mode = (os.getenv('CLAUDE_API_MODE') or 'auto').lower()
    if mode in ('anthropic', 'messages'):
        return 'anthropic'
    if mode in ('invocations', 'chat'):
        return 'invocations'
    return 'auto'


def is_robotic_template_summary(text: str) -> bool:
    return bool(re.match(r'^Unplanned DT % is \d', str(text or '').strip()))


def claude_status() -> dict[str, Any]:
    return {
        'configured': claude_configured(),
        'host': databricks_host() or 'NOT SET',
        'endpoint': claude_endpoint(),
        'api_mode': _claude_api_mode(),
    }


def _top_entries(obj: dict[str, Any], n: int = 5) -> dict[str, Any]:
    if not obj:
        return {}
    entries = list(obj.items())
    if entries and isinstance(entries[0][1], (int, float)):
        entries.sort(key=lambda x: float(x[1]), reverse=True)
        return dict(entries[:n])
    return dict(entries[:n])


def compact_metrics(metrics: dict[str, Any], entity_type: SummaryEntity | None = None) -> dict[str, Any]:
    base: dict[str, Any] = {
        'kpis': metrics.get('kpis'),
        'periods': metrics.get('periods'),
        'period_trend': metrics.get('period_trend'),
        'top_reasons': (metrics.get('reasons') or [])[:5],
        'top_lines': _top_entries(metrics.get('top_lines') or {}, 5),
        'shift_comparison': (metrics.get('shift_comparison') or [])[:3],
    }

    if not entity_type or entity_type == 'overview':
        base['top_sites'] = _top_entries(metrics.get('site_by_period') or {}, 5)
    if not entity_type or entity_type == 'category':
        base['categories'] = _top_entries(metrics.get('category_by_period') or {}, 5)
    if not entity_type or entity_type == 'line':
        base['lines'] = _top_entries(metrics.get('line_by_period') or {}, 5)
    if not entity_type or entity_type == 'dow':
        dow = metrics.get('dow_by_day_week') or {}
        base['dow_sample'] = dict(list(dow.items())[:3])

    return base


def _extract_text(data: dict[str, Any]) -> str:
    choices = data.get('choices')
    if isinstance(choices, list) and choices:
        msg = choices[0].get('message') or {}
        content = msg.get('content')
        if isinstance(content, str) and content.strip():
            return content.strip()

    if isinstance(data.get('output'), str):
        return data['output'].strip()

    if isinstance(data.get('output'), list):
        for block in data['output']:
            if isinstance(block, dict) and block.get('type') == 'message':
                for cb in block.get('content') or []:
                    if isinstance(cb, dict):
                        t = str(cb.get('text') or '').strip()
                        if t:
                            return t

    predictions = data.get('predictions')
    if isinstance(predictions, list) and predictions:
        p = predictions[0]
        if isinstance(p, str):
            return p.strip()
        if isinstance(p, dict):
            if isinstance(p.get('content'), str):
                return p['content'].strip()
            candidates = p.get('candidates')
            if isinstance(candidates, list) and candidates:
                c = candidates[0]
                if isinstance(c, dict) and isinstance(c.get('text'), str):
                    return c['text'].strip()

    return ''


def _parse_narrative(raw: str) -> str:
    text = raw.strip()
    try:
        d = json.loads(text)
        if isinstance(d.get('narrative'), str):
            return d['narrative'].strip()
    except json.JSONDecodeError:
        pass

    cleaned = re.sub(r'```json\s*|```\s*', '', text).strip()
    try:
        d = json.loads(cleaned)
        if isinstance(d.get('narrative'), str):
            return d['narrative'].strip()
    except json.JSONDecodeError:
        pass
    return cleaned[:600]


def _parse_batch_narratives(raw: str) -> dict[SummaryEntity, str]:
    text = re.sub(r'```json\s*|```\s*', '', raw).strip()
    try:
        d = json.loads(text)
        out: dict[SummaryEntity, str] = {}
        for tab in ('overview', 'category', 'line', 'dow', 'reason'):
            v = d.get(tab)
            if isinstance(v, str) and v.strip():
                out[tab] = v.strip()
            elif isinstance(v, dict) and isinstance(v.get('narrative'), str):
                out[tab] = str(v['narrative']).strip()
        if out:
            return out
    except json.JSONDecodeError:
        pass
    return {}


def _invoke_claude_invocations(messages: list[ChatMessage]) -> str:
    h = databricks_host()
    url = f'https://{h}/serving-endpoints/{claude_endpoint()}/invocations'
    body = {
        'messages': messages,
        'max_tokens': int(os.getenv('CLAUDE_MAX_TOKENS') or 800),
        'temperature': float(os.getenv('CLAUDE_TEMPERATURE') or 0.2),
    }

    print(f'[claude] invocations → {claude_endpoint()} ({len(messages)} messages)…', flush=True)
    started = time.time()

    resp = databricks_fetch(
        url,
        method='POST',
        headers={
            'Authorization': f'Bearer {databricks_token()}',
            'Content-Type': 'application/json',
        },
        json_body=body,
    )

    if resp.status_code >= 400:
        err_text = resp.text[:400]
        raise RuntimeError(f'Claude invocations HTTP {resp.status_code}: {err_text}')

    data = resp.json()
    text = _extract_text(data)
    if not text:
        raise RuntimeError('Claude invocations returned empty content')
    elapsed = int((time.time() - started) * 1000)
    print(f'[claude] ✓ invocations {len(text)} chars in {elapsed}ms', flush=True)
    return text


def _invoke_claude_anthropic(messages: list[ChatMessage]) -> str:
    h = databricks_host()
    url = f'https://{h}/serving-endpoints/anthropic/v1/messages'
    system = next((m['content'] for m in messages if m.get('role') == 'system'), None)
    chat_messages = [
        {'role': m['role'], 'content': m['content']}
        for m in messages
        if m.get('role') != 'system'
    ]

    body: dict[str, Any] = {
        'model': claude_endpoint(),
        'max_tokens': int(os.getenv('CLAUDE_MAX_TOKENS') or 800),
        'temperature': float(os.getenv('CLAUDE_TEMPERATURE') or 0.2),
        'messages': chat_messages or [{'role': 'user', 'content': messages[-1]['content']}],
    }
    if system:
        body['system'] = system

    print(f'[claude] anthropic/messages → {claude_endpoint()}…', flush=True)
    started = time.time()

    resp = databricks_fetch(
        url,
        method='POST',
        headers={
            'Authorization': f'Bearer {databricks_token()}',
            'Content-Type': 'application/json',
        },
        json_body=body,
    )

    if resp.status_code >= 400:
        err_text = resp.text[:400]
        raise RuntimeError(f'Claude anthropic HTTP {resp.status_code}: {err_text}')

    data = resp.json()
    blocks = data.get('content')
    if isinstance(blocks, list):
        text = ''.join(str(b.get('text') or '') for b in blocks if isinstance(b, dict)).strip()
    else:
        text = _extract_text(data)
    if not text:
        raise RuntimeError('Claude anthropic returned empty content')
    elapsed = int((time.time() - started) * 1000)
    print(f'[claude] ✓ anthropic {len(text)} chars in {elapsed}ms', flush=True)
    return text


def invoke_claude(messages: list[ChatMessage]) -> str:
    mode = _claude_api_mode()
    if mode == 'anthropic':
        return _invoke_claude_anthropic(messages)
    if mode == 'invocations':
        return _invoke_claude_invocations(messages)

    try:
        return _invoke_claude_invocations(messages)
    except Exception as first:
        msg = str(first)[:120]
        print(f'[claude] invocations failed ({msg}) — trying anthropic/messages', flush=True)
        return _invoke_claude_anthropic(messages)


def get_claude_tab_summary(
    entity_type: SummaryEntity,
    filters: dict[str, Any],
    metrics: dict[str, Any],
) -> dict[str, str]:
    filter_ctx = build_filter_context(filters)
    grounding = build_metrics_grounding_context(metrics, entity_type)

    user = f"""{PROFESSIONAL_VOICE}

Tab focus: {TAB_FOCUS[entity_type]}
Applied filters: {filter_ctx}

{grounding}

Return JSON only: {{"narrative":"your 2-4 sentence executive summary"}}"""

    raw = invoke_claude([{'role': 'user', 'content': user}])
    narrative = _parse_narrative(raw)
    if not narrative:
        raise RuntimeError('Claude returned empty narrative')
    return {'narrative': narrative, 'source': 'claude'}


def get_claude_batch_summaries(
    filters: dict[str, Any],
    metrics: dict[str, Any],
) -> dict[SummaryEntity, str]:
    filter_ctx = build_filter_context(filters)
    grounding = build_metrics_grounding_context(metrics, 'overview')

    user = f"""{PROFESSIONAL_VOICE}

Applied filters: {filter_ctx}

{grounding}

Additional metrics JSON (categories, lines, reasons, shifts):
{json.dumps(compact_metrics(metrics), indent=2)}

Write a distinct executive summary for each KPI tab below.
Return ONLY valid JSON with these keys (each value is a 2-4 sentence string):
{{
  "overview": "...",
  "category": "...",
  "line": "...",
  "dow": "...",
  "reason": "..."
}}"""

    raw = invoke_claude([{'role': 'user', 'content': user}])
    parsed = _parse_batch_narratives(raw)
    tabs: list[SummaryEntity] = ['overview', 'category', 'line', 'dow', 'reason']
    out: dict[SummaryEntity, str] = {}

    for tab in tabs:
        if parsed.get(tab):
            out[tab] = parsed[tab]
        else:
            single = get_claude_tab_summary(tab, filters, metrics)
            out[tab] = single['narrative']

    return out
